NPE when collecting stats

## Description

When calling DockerClient.stats on my containers, an NPE is thrown during the response parsing due to `maxUsage` (along with the rest of the memory stats) missing from the response. The response JSON from docker is definitely missing it and I'm not sure why but I don't really think this is a case where DockerClient should outright throw an NPE. The container is running just fine and docker has no issues with, why should we? 

Unless of course, my docker install is totally busted, then that's understandable.

## How to reproduce

Call DockerClient.stats on a container that doesn't return `maxUsage` (or any memory stats).

## What do you expect

The DockerClient to not throw an NPE.

## What happened instead

The DockerClient threw an NPE.

## Software:

- `docker version`: 17.05.0-ce for client, 17.06.0-ce for server
- Spotify's docker-client version: 8.8.2

## Full backtrace

JSON response from docker:
```json
{  
   "read":"2017-07-28T03:01:05.573009123Z",
   "preread":"2017-07-28T03:01:04.573119273Z",
   "pids_stats":{  

   },
   "blkio_stats":{  
      "io_service_bytes_recursive":[  

      ],
      "io_serviced_recursive":[  

      ],
      "io_queue_recursive":[  

      ],
      "io_service_time_recursive":[  

      ],
      "io_wait_time_recursive":[  

      ],
      "io_merged_recursive":[  

      ],
      "io_time_recursive":[  

      ],
      "sectors_recursive":[  

      ]
   },
   "num_procs":0,
   "storage_stats":{  

   },
   "cpu_stats":{  
      "cpu_usage":{  
         "total_usage":15444440,
         "percpu_usage":[  
            2055152,
            890316,
            331923,
            2011864,
            323424,
            8616546,
            1215215,
            0
         ],
         "usage_in_kernelmode":0,
         "usage_in_usermode":0
      },
      "system_cpu_usage":7257526200000000,
      "online_cpus":8,
      "throttling_data":{  
         "periods":0,
         "throttled_periods":0,
         "throttled_time":0
      }
   },
   "precpu_stats":{  
      "cpu_usage":{  
         "total_usage":15444440,
         "percpu_usage":[  
            2055152,
            890316,
            331923,
            2011864,
            323424,
            8616546,
            1215215,
            0
         ],
         "usage_in_kernelmode":0,
         "usage_in_usermode":0
      },
      "system_cpu_usage":7257518190000000,
      "online_cpus":8,
      "throttling_data":{  
         "periods":0,
         "throttled_periods":0,
         "throttled_time":0
      }
   },
   "memory_stats":{  

   },
   "name":"/fervent_shannon",
   "id":"f861186c9989f670b97ae0e4d3ee5b69be7aed62b677ea3d430bad2b3084621c",
   "networks":{  
      "eth0":{  
         "rx_bytes":648,
         "rx_packets":8,
         "rx_errors":0,
         "rx_dropped":0,
         "tx_bytes":0,
         "tx_packets":0,
         "tx_errors":0,
         "tx_dropped":0
      }
   }
}
```

Full stacktrace:
```text
com.spotify.docker.client.exceptions.DockerRequestException: Request error: GET http://192.168.0.77:2375/containers/f861186c9989f670b97ae0e4d3ee5b69be7aed62b677ea3d430bad2b3084621c/stats?stream=0: 200
	at com.spotify.docker.client.DefaultDockerClient.propagate(DefaultDockerClient.java:2504)
	at com.spotify.docker.client.DefaultDockerClient.request(DefaultDockerClient.java:2443)
	at com.spotify.docker.client.DefaultDockerClient.stats(DefaultDockerClient.java:2128)
	at com.virii.gameseer.supervisor.GameServer.getStats(GameServer.kt:9)
	at com.virii.gameseer.supervisor.views.ServerInfoPanel.setGameServer(ServerInfoPanel.kt:37)
	at com.virii.gameseer.supervisor.views.MainPanel$addServerToList$1.run(MainPanel.kt:41)
	at com.googlecode.lanterna.gui2.ActionListBox$1.run(ActionListBox.java:78)
	at com.googlecode.lanterna.gui2.ActionListBox.handleKeyStroke(ActionListBox.java:100)
	at com.googlecode.lanterna.gui2.AbstractInteractableComponent.handleInput(AbstractInteractableComponent.java:142)
	at com.googlecode.lanterna.gui2.AbstractBasePane.doHandleInput(AbstractBasePane.java:124)
	at com.googlecode.lanterna.gui2.AbstractBasePane.handleInput(AbstractBasePane.java:97)
	at com.googlecode.lanterna.gui2.AbstractWindow.handleInput(AbstractWindow.java:132)
	at com.googlecode.lanterna.gui2.MultiWindowTextGUI.handleInput(MultiWindowTextGUI.java:333)
	at com.googlecode.lanterna.gui2.AbstractTextGUI.processInput(AbstractTextGUI.java:94)
	at com.googlecode.lanterna.gui2.AbstractTextGUIThread.processEventsAndUpdate(AbstractTextGUIThread.java:75)
	at com.googlecode.lanterna.gui2.MultiWindowTextGUI.waitForWindowToClose(MultiWindowTextGUI.java:403)
	at com.googlecode.lanterna.gui2.AbstractWindow.waitUntilClosed(AbstractWindow.java:268)
	at com.googlecode.lanterna.gui2.MultiWindowTextGUI.addWindowAndWait(MultiWindowTextGUI.java:370)
	at com.virii.gameseer.supervisor.SupervisorGUI.<init>(Entry.kt:57)
	at com.virii.gameseer.supervisor.EntryKt.main(Entry.kt:21)
Caused by: javax.ws.rs.client.ResponseProcessingException: com.fasterxml.jackson.databind.JsonMappingException: Can not construct instance of com.spotify.docker.client.messages.MemoryStats, problem: Null maxUsage
 at [Source: org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream@66285b3c; line: 1, column: 969] (through reference chain: com.spotify.docker.client.messages.ContainerStats["memory_stats"])
	at org.glassfish.jersey.client.JerseyInvocation.translate(JerseyInvocation.java:806)
	at org.glassfish.jersey.client.JerseyInvocation.access$700(JerseyInvocation.java:92)
	at org.glassfish.jersey.client.JerseyInvocation$5.completed(JerseyInvocation.java:773)
	at org.glassfish.jersey.client.ClientRuntime.processResponse(ClientRuntime.java:198)
	at org.glassfish.jersey.client.ClientRuntime.access$300(ClientRuntime.java:79)
	at org.glassfish.jersey.client.ClientRuntime$2.run(ClientRuntime.java:180)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:340)
	at org.glassfish.jersey.client.ClientRuntime$3.run(ClientRuntime.java:210)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Can not construct instance of com.spotify.docker.client.messages.MemoryStats, problem: Null maxUsage
 at [Source: org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream@66285b3c; line: 1, column: 969] (through reference chain: com.spotify.docker.client.messages.ContainerStats["memory_stats"])
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:277)
	at com.fasterxml.jackson.databind.DeserializationContext.instantiationException(DeserializationContext.java:1441)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.wrapAsJsonMappingException(StdValueInstantiator.java:476)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.rewrapCtorProblem(StdValueInstantiator.java:495)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromObjectWith(StdValueInstantiator.java:276)
	at com.fasterxml.jackson.databind.deser.ValueInstantiator.createFromObjectWith(ValueInstantiator.java:228)
	at com.fasterxml.jackson.databind.deser.impl.PropertyBasedCreator.build(PropertyBasedCreator.java:138)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:471)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1191)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:148)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:504)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:511)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:400)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1191)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:148)
	at com.fasterxml.jackson.databind.ObjectReader._bind(ObjectReader.java:1583)
	at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:964)
	at com.fasterxml.jackson.jaxrs.base.ProviderBase.readFrom(ProviderBase.java:815)
	at org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$TerminalReaderInterceptor.invokeReadFrom(ReaderInterceptorExecutor.java:256)
	at org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$TerminalReaderInterceptor.aroundReadFrom(ReaderInterceptorExecutor.java:235)
	at org.glassfish.jersey.message.internal.ReaderInterceptorExecutor.proceed(ReaderInterceptorExecutor.java:155)
	at org.glassfish.jersey.message.internal.MessageBodyFactory.readFrom(MessageBodyFactory.java:1085)
	at org.glassfish.jersey.message.internal.InboundMessageContext.readEntity(InboundMessageContext.java:874)
	at org.glassfish.jersey.message.internal.InboundMessageContext.readEntity(InboundMessageContext.java:808)
	at org.glassfish.jersey.client.ClientResponse.readEntity(ClientResponse.java:326)
	at org.glassfish.jersey.client.JerseyInvocation.translate(JerseyInvocation.java:803)
	... 17 more
Caused by: java.lang.NullPointerException: Null maxUsage
	at com.spotify.docker.client.messages.AutoValue_MemoryStats.<init>(AutoValue_MemoryStats.java:25)
	at com.spotify.docker.client.messages.MemoryStats.create(MemoryStats.java:63)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.fasterxml.jackson.databind.introspect.AnnotatedMethod.call(AnnotatedMethod.java:117)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromObjectWith(StdValueInstantiator.java:274)
	... 40 more
```

--- a/src/main/java/com/spotify/docker/client/messages/MemoryStats.java
+++ b/src/main/java/com/spotify/docker/client/messages/MemoryStats.java
@@ -39,9 +39,11 @@ public abstract class MemoryStats {
   @JsonProperty("stats")
   public abstract Stats stats();
 
+  @Nullable
   @JsonProperty("max_usage")
   public abstract Long maxUsage();
 
+  @Nullable
   @JsonProperty("usage")
   public abstract Long usage();
 
@@ -49,6 +51,7 @@ public abstract class MemoryStats {
   @JsonProperty("failcnt")
   public abstract Long failcnt();
 
+  @Nullable
   @JsonProperty("limit")
   public abstract Long limit();
 

--- a/src/test/java/com/spotify/docker/client/messages/ContainerStatsTest.java
+++ b/src/test/java/com/spotify/docker/client/messages/ContainerStatsTest.java
@@ -47,4 +47,9 @@ public class ContainerStatsTest {
     objectMapper.readValue(fixture("fixtures/1.26/containerStats.json"), ContainerStats.class);
   }
 
+  @Test
+  public void test1_30() throws Exception {
+    objectMapper.readValue(fixture("fixtures/1.30/containerStats.json"), ContainerStats.class);
+  }
+
 }

--- a/src/test/resources/fixtures/1.30/containerStats.json
+++ b/src/test/resources/fixtures/1.30/containerStats.json
@@ -1,0 +1,102 @@
+{
+  "read":"2017-07-28T03:01:05.573009123Z",
+  "preread":"2017-07-28T03:01:04.573119273Z",
+  "pids_stats":{
+
+  },
+  "blkio_stats":{
+    "io_service_bytes_recursive":[
+
+    ],
+    "io_serviced_recursive":[
+
+    ],
+    "io_queue_recursive":[
+
+    ],
+    "io_service_time_recursive":[
+
+    ],
+    "io_wait_time_recursive":[
+
+    ],
+    "io_merged_recursive":[
+
+    ],
+    "io_time_recursive":[
+
+    ],
+    "sectors_recursive":[
+
+    ]
+  },
+  "num_procs":0,
+  "storage_stats":{
+
+  },
+  "cpu_stats":{
+    "cpu_usage":{
+      "total_usage":15444440,
+      "percpu_usage":[
+        2055152,
+        890316,
+        331923,
+        2011864,
+        323424,
+        8616546,
+        1215215,
+        0
+      ],
+      "usage_in_kernelmode":0,
+      "usage_in_usermode":0
+    },
+    "system_cpu_usage":7257526200000000,
+    "online_cpus":8,
+    "throttling_data":{
+      "periods":0,
+      "throttled_periods":0,
+      "throttled_time":0
+    }
+  },
+  "precpu_stats":{
+    "cpu_usage":{
+      "total_usage":15444440,
+      "percpu_usage":[
+        2055152,
+        890316,
+        331923,
+        2011864,
+        323424,
+        8616546,
+        1215215,
+        0
+      ],
+      "usage_in_kernelmode":0,
+      "usage_in_usermode":0
+    },
+    "system_cpu_usage":7257518190000000,
+    "online_cpus":8,
+    "throttling_data":{
+      "periods":0,
+      "throttled_periods":0,
+      "throttled_time":0
+    }
+  },
+  "memory_stats":{
+
+  },
+  "name":"/fervent_shannon",
+  "id":"f861186c9989f670b97ae0e4d3ee5b69be7aed62b677ea3d430bad2b3084621c",
+  "networks":{
+    "eth0":{
+      "rx_bytes":648,
+      "rx_packets":8,
+      "rx_errors":0,
+      "rx_dropped":0,
+      "tx_bytes":0,
+      "tx_packets":0,
+      "tx_errors":0,
+      "tx_dropped":0
+    }
+  }
+}

